memcpy intrinsic tests failing on ARM64 with LLVM-13

I am seeing the following errors with rellic using LLVM-13 on ARM64 machine (from cxx-common run here https://github.com/lifting-bits/cxx-common/runs/5254913519?check_suite_focus=true#step:15:171)

```
FAIL: test_byval_struct (__main__.TestRoundtrip)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "../scripts/roundtrip.py", line 120, in test
    roundtrip(self, args.rellic, path, args.clang, args.timeout, args.translate_only)
  File "../scripts/roundtrip.py", line 92, in roundtrip
    compile(self, clang, rt_c, out2, timeout, ["-Wno-everything"])
  File "../scripts/roundtrip.py", line 44, in compile
    self.assertEqual(
AssertionError: 204 != 0 : errors or warnings during compilation: /usr/bin/ld: /tmp/rt-0482d1.o: in function `main':
rt.c:(.text+0xbc): undefined reference to `llvm_memcpy_p0i8_p0i8_i64'
clang-13: error: linker command failed with exit code 1 (use -v to see invocation)


======================================================================
FAIL: test_issue_183_literal_structs (__main__.TestRoundtrip)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "../scripts/roundtrip.py", line 120, in test
    roundtrip(self, args.rellic, path, args.clang, args.timeout, args.translate_only)
  File "../scripts/roundtrip.py", line 81, in roundtrip
    decompile(self, rellic, rt_bc, rt_c, timeout)
  File "../scripts/roundtrip.py", line 59, in decompile
    self.assertEqual(p.returncode, 0, "rellic-decomp failure: %s" % p.stderr)
AssertionError: -6 != 0 : rellic-decomp failure: error: array type 'unsigned long [2]' is not assignable
F0219 03:17:10.542593 25774 ASTBuilder.cpp:355] Check failed: er.isUsable() 
```

@pgoodman says "we should be lifting those intrinsics to calls to memcpy"

--- a/include/rellic/AST/ASTBuilder.h
+++ b/include/rellic/AST/ASTBuilder.h
@@ -8,6 +8,7 @@
 
 #pragma once
 
+#include <clang/Basic/Builtins.h>
 #include <clang/Frontend/ASTUnit.h>
 
 #include <string>
@@ -271,6 +272,9 @@ class ASTBuilder {
                               std::vector<clang::Expr *> &args) {
     return CreateCall(CreateDeclRef(func), args);
   }
+  // Intrinsic call
+  clang::Expr *CreateBuiltinCall(clang::Builtin::ID builtin,
+                                 std::vector<clang::Expr *> &args);
   // Structure field access
   clang::MemberExpr *CreateFieldAcc(clang::Expr *base, clang::FieldDecl *field,
                                     bool is_arrow);

--- a/include/rellic/AST/IRToASTVisitor.h
+++ b/include/rellic/AST/IRToASTVisitor.h
@@ -13,6 +13,7 @@
 #include <llvm/IR/Argument.h>
 #include <llvm/IR/InlineAsm.h>
 #include <llvm/IR/InstVisitor.h>
+#include <llvm/IR/IntrinsicInst.h>
 #include <llvm/IR/Operator.h>
 #include <llvm/IR/Value.h>
 #include <rellic/AST/Util.h>
@@ -68,6 +69,10 @@ class IRToASTVisitor : public llvm::InstVisitor<IRToASTVisitor> {
   void VisitFunctionDecl(llvm::Function &func);
   void VisitArgument(llvm::Argument &arg);
 
+  void visitMemCpyInst(llvm::MemCpyInst &inst);
+  void visitMemCpyInlineInst(llvm::MemCpyInlineInst &inst);
+  void visitAnyMemMoveInst(llvm::AnyMemMoveInst &inst);
+  void visitAnyMemSetInst(llvm::AnyMemSetInst &inst);
   void visitIntrinsicInst(llvm::IntrinsicInst &inst);
   void visitCallInst(llvm::CallInst &inst);
   void visitGetElementPtrInst(llvm::GetElementPtrInst &inst);

--- a/lib/AST/ASTBuilder.cpp
+++ b/lib/AST/ASTBuilder.cpp
@@ -8,7 +8,11 @@
 
 #include "rellic/AST/ASTBuilder.h"
 
+#include <clang/AST/ASTContext.h>
+#include <clang/AST/Decl.h>
+#include <clang/Basic/SourceLocation.h>
 #include <clang/Basic/TargetInfo.h>
+#include <clang/Sema/Lookup.h>
 #include <clang/Sema/Sema.h>
 #include <gflags/gflags.h>
 #include <glog/logging.h>
@@ -389,6 +393,22 @@ clang::CallExpr *ASTBuilder::CreateCall(clang::Expr *callee,
   return er.getAs<clang::CallExpr>();
 }
 
+clang::Expr *ASTBuilder::CreateBuiltinCall(clang::Builtin::ID builtin,
+                                           std::vector<clang::Expr *> &args) {
+  auto name{ctx.BuiltinInfo.getName(builtin)};
+  clang::SourceLocation loc;
+  clang::LookupResult R(sema, &ctx.Idents.get(name), loc,
+                        clang::Sema::LookupOrdinaryName);
+  clang::Sema::LookupNameKind NameKind = R.getLookupKind();
+  auto II{R.getLookupName().getAsIdentifierInfo()};
+  clang::ASTContext::GetBuiltinTypeError error;
+  auto ty{ctx.GetBuiltinType(builtin, error)};
+  CHECK(!error);
+  auto decl{sema.CreateBuiltin(II, ty, builtin, loc)};
+
+  return CreateCall(decl, args);
+}
+
 clang::MemberExpr *ASTBuilder::CreateFieldAcc(clang::Expr *base,
                                               clang::FieldDecl *field,
                                               bool is_arrow) {
@@ -459,8 +479,9 @@ clang::WhileStmt *ASTBuilder::CreateWhile(clang::Expr *cond,
 
 clang::DoStmt *ASTBuilder::CreateDo(clang::Expr *cond, clang::Stmt *body) {
   // auto sr{sema.ActOnDoStmt(clang::SourceLocation(), body,
-  //                          clang::SourceLocation(), clang::SourceLocation(),
-  //                          cond, clang::SourceLocation())};
+  //                          clang::SourceLocation(),
+  //                          clang::SourceLocation(), cond,
+  //                          clang::SourceLocation())};
   // CHECK(sr.isUsable());
   // return sr.getAs<clang::DoStmt>();
   CHECK(cond != nullptr) << "Should not be null in CreateDo.";

--- a/lib/AST/IRToASTVisitor.cpp
+++ b/lib/AST/IRToASTVisitor.cpp
@@ -514,7 +514,7 @@ void IRToASTVisitor::visitMemCpyInst(llvm::MemCpyInst &inst) {
     args.push_back(GetOperandExpr(arg));
   }
 
-  callstmt = ast.CreateIntrinsicCall(clang::Builtin::BI__builtin_memcpy, args);
+  callstmt = ast.CreateBuiltinCall(clang::Builtin::BI__builtin_memcpy, args);
 }
 
 void IRToASTVisitor::visitMemCpyInlineInst(llvm::MemCpyInlineInst &inst) {
@@ -530,7 +530,7 @@ void IRToASTVisitor::visitMemCpyInlineInst(llvm::MemCpyInlineInst &inst) {
     args.push_back(GetOperandExpr(arg));
   }
 
-  callstmt = ast.CreateIntrinsicCall(clang::Builtin::BI__builtin_memcpy, args);
+  callstmt = ast.CreateBuiltinCall(clang::Builtin::BI__builtin_memcpy, args);
 }
 
 void IRToASTVisitor::visitAnyMemMoveInst(llvm::AnyMemMoveInst &inst) {
@@ -546,7 +546,7 @@ void IRToASTVisitor::visitAnyMemMoveInst(llvm::AnyMemMoveInst &inst) {
     args.push_back(GetOperandExpr(arg));
   }
 
-  callstmt = ast.CreateIntrinsicCall(clang::Builtin::BI__builtin_memmove, args);
+  callstmt = ast.CreateBuiltinCall(clang::Builtin::BI__builtin_memmove, args);
 }
 
 void IRToASTVisitor::visitAnyMemSetInst(llvm::AnyMemSetInst &inst) {
@@ -562,7 +562,7 @@ void IRToASTVisitor::visitAnyMemSetInst(llvm::AnyMemSetInst &inst) {
     args.push_back(GetOperandExpr(arg));
   }
 
-  callstmt = ast.CreateIntrinsicCall(clang::Builtin::BI__builtin_memset, args);
+  callstmt = ast.CreateBuiltinCall(clang::Builtin::BI__builtin_memset, args);
 }
 
 void IRToASTVisitor::visitIntrinsicInst(llvm::IntrinsicInst &inst) {

--- a/lib/AST/IRToASTVisitor.cpp
+++ b/lib/AST/IRToASTVisitor.cpp
@@ -6,7 +6,9 @@
  * the LICENSE file found in the root directory of this source tree.
  */
 
+#include <clang/Basic/Builtins.h>
 #include <llvm/IR/Instructions.h>
+#include <llvm/IR/IntrinsicInst.h>
 #define GOOGLE_STRIP_LOG 1
 
 #include <gflags/gflags.h>
@@ -497,6 +499,70 @@ void IRToASTVisitor::VisitFunctionDecl(llvm::Function &func) {
   }
 
   decl->getAsFunction()->setParams(params);
+}
+
+void IRToASTVisitor::visitMemCpyInst(llvm::MemCpyInst &inst) {
+  DLOG(INFO) << "visitMemCpyInlineInst: " << LLVMThingToString(&inst);
+  auto &callstmt{stmts[&inst]};
+  if (callstmt) {
+    return;
+  }
+
+  std::vector<clang::Expr *> args;
+  for (auto i{0U}; i < 3; ++i) {
+    auto &arg{inst.getArgOperandUse(i)};
+    args.push_back(GetOperandExpr(arg));
+  }
+
+  callstmt = ast.CreateIntrinsicCall(clang::Builtin::BI__builtin_memcpy, args);
+}
+
+void IRToASTVisitor::visitMemCpyInlineInst(llvm::MemCpyInlineInst &inst) {
+  DLOG(INFO) << "visitMemCpyInlineInst: " << LLVMThingToString(&inst);
+  auto &callstmt{stmts[&inst]};
+  if (callstmt) {
+    return;
+  }
+
+  std::vector<clang::Expr *> args;
+  for (auto i{0U}; i < 3; ++i) {
+    auto &arg{inst.getArgOperandUse(i)};
+    args.push_back(GetOperandExpr(arg));
+  }
+
+  callstmt = ast.CreateIntrinsicCall(clang::Builtin::BI__builtin_memcpy, args);
+}
+
+void IRToASTVisitor::visitAnyMemMoveInst(llvm::AnyMemMoveInst &inst) {
+  DLOG(INFO) << "visitAnyMemMoveInst: " << LLVMThingToString(&inst);
+  auto &callstmt{stmts[&inst]};
+  if (callstmt) {
+    return;
+  }
+
+  std::vector<clang::Expr *> args;
+  for (auto i{0U}; i < 3; ++i) {
+    auto &arg{inst.getArgOperandUse(i)};
+    args.push_back(GetOperandExpr(arg));
+  }
+
+  callstmt = ast.CreateIntrinsicCall(clang::Builtin::BI__builtin_memmove, args);
+}
+
+void IRToASTVisitor::visitAnyMemSetInst(llvm::AnyMemSetInst &inst) {
+  DLOG(INFO) << "visitAnyMemSetInst: " << LLVMThingToString(&inst);
+  auto &callstmt{stmts[&inst]};
+  if (callstmt) {
+    return;
+  }
+
+  std::vector<clang::Expr *> args;
+  for (auto i{0U}; i < 3; ++i) {
+    auto &arg{inst.getArgOperandUse(i)};
+    args.push_back(GetOperandExpr(arg));
+  }
+
+  callstmt = ast.CreateIntrinsicCall(clang::Builtin::BI__builtin_memset, args);
 }
 
 void IRToASTVisitor::visitIntrinsicInst(llvm::IntrinsicInst &inst) {

--- a/lib/AST/IRToASTVisitor.cpp
+++ b/lib/AST/IRToASTVisitor.cpp
@@ -502,7 +502,7 @@ void IRToASTVisitor::VisitFunctionDecl(llvm::Function &func) {
 }
 
 void IRToASTVisitor::visitMemCpyInst(llvm::MemCpyInst &inst) {
-  DLOG(INFO) << "visitMemCpyInlineInst: " << LLVMThingToString(&inst);
+  DLOG(INFO) << "visitMemCpyInst: " << LLVMThingToString(&inst);
   auto &callstmt{stmts[&inst]};
   if (callstmt) {
     return;

